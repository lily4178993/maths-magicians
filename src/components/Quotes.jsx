import React, { useEffect, useState } from 'react';

const Quotes = () => {
  const [quotes, setQuotes] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState('');

  const baseURL = process.env.REACT_APP_QUOTES_URL;
  const apiKey = process.env.REACT_APP_QUOTES_KEY;
  const category = process.env.REACT_APP_QUOTES_CATEGORY;

  useEffect(() => {
    const fetchQuotes = async () => {
      try {
        const response = await fetch(baseURL + category, {
          headers: {
            'X-Api-Key': apiKey,
          },
          contentType: 'application/json',
          'X-Content-Type-Options': 'nosniff',
        });

        if (!response.ok) {
          throw new Error(
            `Failed to fetch quotes - Status: ${response.status}`,
          );
        }

        const data = await response.json();
        setQuotes(data);
      } catch (error) {
        setError(error.message);
      } finally {
        setLoading(false);
      }
    };

    if (quotes.length === 0) {
      fetchQuotes();
    }
  }, [apiKey, baseURL, category, quotes.length]);

  return (
    <>
      {loading && <div className="loading-message">Loading...</div>}
      {error && <div className="error-message">{error}</div>}
      {quotes.map((item) => (
        <p key={1} className="quote">
          <i>
            <b>
              <cite>{item.quote}</cite>
              <br />
              <br />
              {item.author}
            </b>
          </i>
        </p>
      ))}
    </>
  );
};

export default Quotes;
