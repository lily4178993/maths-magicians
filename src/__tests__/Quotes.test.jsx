import React from 'react';
import { render, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Quotes from '../components/Quotes';

// Mock environment variables
beforeEach(() => {
  process.env.REACT_APP_QUOTES_URL = 'mocked-url';
  process.env.REACT_APP_QUOTES_API_KEY = 'mocked-api-key';
  process.env.REACT_APP_QUOTES_CATEGORY = 'mocked-category';
});

afterEach(() => {
  delete process.env.REACT_APP_QUOTES_URL;
  delete process.env.REACT_APP_QUOTES_API_KEY;
  delete process.env.REACT_APP_QUOTES_CATEGORY;
});

// Mock fetch function
const mockFetch = (data, status = 200) => jest.fn().mockResolvedValue({
  ok: status < 400,
  status,
  json: () => Promise.resolve(data),
});

describe('Quotes component', () => {
  it('matches snapshot while fetching quotes', async () => {
    global.fetch = mockFetch([]);
    const { asFragment } = render(<Quotes />);

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('matches snapshot when fetching quotes is successful', async () => {
    const quotes = [
      { quote: 'Quote 1', author: 'Author 1' },
      { quote: 'Quote 2', author: 'Author 2' },
    ];

    global.fetch = mockFetch(quotes);
    const { asFragment } = render(<Quotes />);

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
