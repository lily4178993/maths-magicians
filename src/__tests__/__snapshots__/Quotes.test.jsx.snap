// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quotes component matches snapshot when fetching quotes is successful 1`] = `
<DocumentFragment>
  <div
    class="loading-message"
  >
    Loading...
  </div>
</DocumentFragment>
`;

exports[`Quotes component matches snapshot while fetching quotes 1`] = `
<DocumentFragment>
  <div
    class="loading-message"
  >
    Loading...
  </div>
</DocumentFragment>
`;
